### PR TITLE
Sign in with apple

### PR DESCRIPTION
**Pull request by geraldnolan**
This PR changes:

- Adds ability to sign in with apple  #169 

To test you must secure your local development server. (**next-auth-example**).  Apple doesn't allow for localhost in domains or subdomains.

I refactored the code and removed the option "oauth-apple"

How to:
[https://medium.com/@anMagpie/secure-your-local-development-server-with-https-next-js-81ac6b8b3d68](https://medium.com/@anMagpie/secure-your-local-development-server-with-https-next-js-81ac6b8b3d68)

Edit your **hostfile** and create a fake domain: 

127.0.0.1 dev.iaincollins.io

This documentation could be simplified but for the most part I followed: (Creating Apple Services IDs)
[https://developer.okta.com/blog/2019/06/04/what-the-heck-is-sign-in-with-apple](https://developer.okta.com/blog/2019/06/04/what-the-heck-is-sign-in-with-apple)

The tricky part after that is creating the secret.  The two things that tripped me up were:

**TeamID** and the **KeyID**.

The TeamID is located on the top right after logging in.
The KeyID is located after you create the Key look for before you download the k8 file. 

```
let privateKey=`-----BEGIN PRIVATE KEY-----
-----END PRIVATE KEY-----`;

let exp = Math.floor(Date.now() / 1000) + ( 86400 * 180 ); // Make it expire within 6 months
  let keyid = "INSERT_KEYID"; //KEYID

  const claims = {
    iss: "INSERT_TEAMID", //TeamID
    iat: Math.floor(Date.now() / 1000),
    exp,
    aud: 'https://appleid.apple.com',
    sub: process.env.APPLE_ID
  };
  // Sign the claims using the private key
  var clientSecret = jwt.sign(claims, privateKey, {
    algorithm: 'ES256',
    keyid
  });

 console.log(clientSecret);

```

Let me know if you have any questions

### CODE REVIEW

**Comment by iaincollins** [comment]
@geraldnolan Oh wow, this looks amazing!

I'm taking a small break and won't be able to try it out myself Thursday or Friday but this looks great, and I can't see anything I'd do differently. 

Thank for for wading through that sketchy oAuth code to do this, the existing code really needs refactoring as it has been hacked around a but and isn't the easiest thing to work with (without dumping console.log statements all over the place) 🙈.

Thank you for the detailed instructions! I'll let you know how I get on!

@LoriKarikari Reading this, am thinking of changing my mind and that maybe having a more detailed document page per provider *is* a good idea for *some* providers after all? I think the table is perfect for 99% of cases, but looking at this case, I'm grateful for the extra info!

**Comment by LoriKarikari** [comment]
@iaincollins yeah something like this needs detailed instructions for sure. But the other 99% is just register your app, get your tokens and add your callback url 😂 so i wouldn't document those

**Comment by iaincollins** [comment]
Thanks again @geraldnolan !

_tl;dr I can't test this end to end right now, but I'm going to merge it in because it looks good as far as I can tell!_

## Testing

* I followed the documentation and those links and the extra info you provided was great!
* Unfortunately, my ADC account has expired, so I can't create an application in ADC
* However, I checked out the branch and I do see the expected behaviour when I use Safari, even though the journey cannot complete (because I didn't provide valid credentials to the provider).

  *It seems you need a valid Client ID to even this prompt in other browsers as Apple fallback to displaying a webpage that checks it; only Safari has a built in native dialog).*

Screenshot 2020-05-27 at 14 02 22
Screenshot 2020-05-27 at 14 02 37

* I also tested out the changes don't cause any other problems with other providers.

  I think the approach of hardcoding some logic here specific to Apple is fine in this context.

  We can always refactor it a bit `src/server/lib/oauth/callback.js` is cleaned up in future. The comments there are very helpful. Including `jsw-decode` is good too, as I am going to pull in a lib for that anyway when I come to add JWT token support, so I'm sure it won't just be for the Apple provider.

I'm going to merge this in now, and will update the documentation for this provider, and re-use the information you have provided here when I do that.

## Questions

* Do you still specify `clientId` and `clientSecret` when using the Apple provider?
* `clientId` should be the value in ADC for `client_id` - is that correct?
* `clientSecret` only needs to be generated once ever, using the example code - is that correct?
* What is `process.env.APPLE_ID` in that context (is that an email address?)

Thank you!

**Review by iaincollins — APPROVED** [review]
Approved: See comments and questions in conversation of #174 (above)

**Comment by geraldnolan** [comment]
@iaincollins I thought you were taking a break? :smiley:

**Do you still specify clientId and clientSecret when using the Apple provider?**

Yes, In the case I generated secret outside the app and placed in the env file.  We may want more sophistication and help the user generate the clientSecret.  I didn't put that code in **next-auth** cause I thought the end user wanted some control?

```
Providers.Apple({
      clientId: process.env.APPLE_ID,
      clientSecret: process.env.APPLE_SECRET
    }),
```

**clientId should be the value in ADC for client_id - is that correct?**

From the example appove it should be: (env_file) 

```
TWITTER_ID=
TWITTER_SECRET=
GITHUB_ID=
GITHUB_SECRET=
TWITCH_ID=
TWITCH_SECRET=
APPLE_ID=dev.iaincollins.io
APPLE_SECRET=**GENERATED_JWT_TOKEN_FROM_CODE_ABOVE**
```

** its the servicesid from the developer portal

**clientSecret only needs to be generated once ever, using the example code - is that correct?**

The clientSecret will need to generated once the token expires.  I set it to 6 months. (Maximum time for Apple)

**What is process.env.APPLE_ID in that context (is that an email address?)**
from env file: dev.iaincollins.io

Hopefully I answered those in ways that makes sense. Shoot me another message if I need to clarify anything else.

**Comment by iaincollins** [comment]
Thank you, that's very clear!

I just have one follow up question!

RE: Secret generation

I'm considering writing a little command line function to come with the library, to help with things like setting up database schemas the first time and checking the current schema, and was thinking the code snippet you provided might be useful to in that context

e.g. so folks do something like this:

 `npx next-auth secret apple {APPLE_ID} {KEY_ID} {TEAM_ID}`

But that lead me to wonder, do you think it would be appropriate to consider generate the secret automatically at run time instead, to make this easier?

e.g. The behaviour could be refactored to allow `clientSecret` to be an object (other than a string) and running a callback on it if defined.

e.g. it could look like this in the front end:

```javascript
Providers.Apple({
  clientId: process.env.APPLE_ID,
  clientSecret: { 
    clientId: process.env.APPLE_ID,
    teamId: process.env.APPLE_TEAM_ID,
    privateKey: process.env.APPLE_PRIVATE_KEY,
    keyId: process.env.APPLE_KEY_ID,
  }
}),
```

And in `src/providers/apple.js` it could have a pre-defined clientSecret function like this:

```javascript
  ...
  clientSecret: ({appleId, keyId, teamId, privateKey}) => {
      const claims = {
        iss: teamId,
        iat: Math.floor(Date.now() / 1000),
        exp: Math.floor(Date.now() / 1000) + ( 86400 * 180 ), // 6 months
        aud: 'https://appleid.apple.com',
        sub: appleId
      }
    
      return jwt.sign(claims, privateKey, {
        algorithm: 'ES256',
        keyid
      })
  })
```

* Alternatively, client secret in the client could look something like this, and the function in `src/providers/apple.js` could be called something else, like `clientSecretCallback`:

```javascript
Providers.Apple({
  clientId: process.env.APPLE_ID,
  clientSecret: { 
    appleId: process.env.APPLE_ID,
    keyId: process.env.APPLE_KEY_ID,
    teamId: process.env.APPLE_TEAM_ID,
    privateKey: process.env.APPLE_PRIVATE_KEY
  }
}),
```

Do you think an approach like this would make sense, or not really?

If it's computationally expensive, we could probably have it invoked once at app start up and then re-used on subsequent calls (as even a Serverless function stays memory resident and doesn't run from scratch every time it's called).

**Comment by iaincollins** [comment]
> @iaincollins I thought you were taking a break? 😃

Also: LOL 🙈

**Comment by geraldnolan** [comment]
```
clientSecret: ({appleId, keyId, teamId, privateKey}) => {
      const claims = {
        iss: teamId,
        iat: Math.floor(Date.now() / 1000),
        exp: Math.floor(Date.now() / 1000) + ( 86400 * 180 ), // 6 months
        aud: 'https://appleid.apple.com',
        sub: appleId
      }
    
      return jwt.sign(claims, privateKey, {
        algorithm: 'ES256',
        keyid
      })
  })
```

@iaincollins I like your thinking there.  You will have less issues being raised.

One issue I found is trying to put the private key in as env variable.  It's a little bit tricky to put it in a single line.

process.env.APPLE_PRIVATE_KEY  might have to be a location path on disk?

The exp claim should be configurable as well.

**Comment by iaincollins** [comment]
@geraldnolan Thank you, that's very helpful!

I agree, I wondered about the difficulty of putting a key in an environment variable too.

At least with `now.sh` command line tool - and with the GitHub secrets UI - multiline environment variables are fairly easy, as the former can read them from stdin and the GitHub secrets UI lets you paste them, but it's much more alkward in an `.env` file, so maybe the docs ought to show how to use it with `fs.readFileSync()` as well for easier local development.

I really appreciate this; this is a wonderful feature to have!

If I do any refactoring of it as above I'll test it as best I can, but will tag you in the update so you can see if/when it goes out so you have a heads up!
